sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusProcessorClient.java

```diff
@@ -376,10 +376,19 @@ public void onComplete() {
             };
         }
 
-        receiverClient.receiveMessagesWithContext()
-            .parallel(processorOptions.getMaxConcurrentCalls(), 1)
-            .runOn(Schedulers.boundedElastic(), 1)
-            .subscribe(subscribers);
+        if (processorOptions.getMaxConcurrentCalls() > 1) {
+            receiverClient.receiveMessagesWithContext()
+                .parallel(processorOptions.getMaxConcurrentCalls(), 1)
+                .runOn(Schedulers.boundedElastic(), 1)
+                .subscribe(subscribers);
+        } else {
+            // There is no need to incur the above cost of the parallel operator
+            // and associated thread hopping (via runOn) when the max-concurrent-call
+            // is default (one), the low-level receiver i.e. receiverClient object
+            // already publisheOn a Bounded-Elastic thread, which is good enough.
+            receiverClient.receiveMessagesWithContext()
+                .subscribe(subscribers[0]);
```

Comment:
Though the focus was on removing thread hopping, a "good" side effect of this is - the removal of undesired prefetching.

Often, applications combine the default (i.e., 1) max-concurrent-calls with prefetch=0, disabling prefetching.

```java
ServiceBusProcessorClient receiver = new ServiceBusClientBuilder()
    .connectionString(connectionString)
    .processor()
    .disableAutoComplete()
    .maxAutoLockRenewDuration(Duration.ofSeconds(60))
    .queueName(queueName)
    .prefetchCount(0)
    .maxConcurrentCalls(1)
    .processMessage(message -> {
        System.out.println(Thread.currentThread().getName() + ":" + "Processing Message [Start]:" + message.getMessage().getMessageId());
        try {
            TimeUnit.SECONDS.sleep(5);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        System.out.println(Thread.currentThread().getName() + ":" + "Processing Message [End]:" + message.getMessage().getMessageId());
    })
    .processError(errorContext -> {
        errorContext.getException().printStackTrace();
    })
    .buildProcessorClient();
```

But due to earlier parallel()+runOn() in max-concurrent-calls=1 code path, there used to be undesired prefetching; 

### below is an example log produced by old code -

<img width="1626" alt="old-code-prefetching" src="https://user-images.githubusercontent.com/1471612/169708070-a8b06a01-db5e-4421-9531-d42fd69b7e6e.png">

Line2 places `credit_a(1)` upon starting the processor, and the respective `message_a` is delivered in line9. 

The `message_a` en-queued to parallel()+runOn(), resulting in back pressure and event-loop-thread places `credit_b(1)` in line12, then bounded-elastic-thread calls processor handler in line13 with `message_a`. The handler takes 5 sec to process a message.

While the handler processes `message_a` - 
1. the `message_b` arrives and gets cached, line14.
2. another `credit_c(1)` is placed, corresponding `message_c` came and cached as well, line17 and 18

line20 finishes the processing of `message_a`. Before calling the processor handler with `message_b` in line24, we can also see the placement of `credit_d(1)` in line23.

### Let's compare above log with the log produced by the new code

Here can see there is no undesired credit placement (hence no prefetching). There is a serial ordering, i.e., a credit(1) placement for the next message happens only after the processor handler completes the 5 sec processing of the current message.

<img width="1626" alt="new-code-prefetching" src="https://user-images.githubusercontent.com/1471612/169708223-12b28856-c051-4836-99fe-ad2fd71dc2ae.png">

Reply:
nit; `addCredits(1)` appearing in the logs is a temporary dev-time logging I added to confirm the credit flow.

Reply:
naive question:

I assume we use `Flux.parallel()` because some callback from user could run as sync code block, and hence `boundedElastic` scheduler.

Does the changed code also run on a scheduler that allows sync code (it probably is, but should we comment this precondition here, if indeed is required?).

Reply:
Yes modified code already run on bounded elastic, I did comment the new code, maybe not clear enough, I’ll reword it.

Reply:
updated the code comment.